controller_manager:
  ros__parameters:
    update_rate: 1000  # Hz

    gravity_compensation_example_controller:
      type: franka_example_controllers/GravityCompensationExampleController

    joint_impedance_example_controller:
      type: franka_example_controllers/JointImpedanceExampleController

    cartesian_impedance_controller:
      type: franka_example_controllers/CartesianImpedanceController

    joint_position_example_controller:
      type: franka_example_controllers/JointPositionExampleController

    joint_velocity_example_controller:
      type: franka_example_controllers/JointVelocityExampleController
    
    cartesian_velocity_example_controller:
      type: franka_example_controllers/CartesianVelocityExampleController

    move_to_start_example_controller:
      type: franka_example_controllers/MoveToStartExampleController

    joint_trajectory_controller:
      type: joint_effort_trajectory_controller/JointTrajectoryController

    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster

    franka_robot_state_broadcaster:
      type: franka_robot_state_broadcaster/FrankaRobotStateBroadcaster

    model_example_controller:
      type: franka_example_controllers/ModelExampleController

    # Cartesian Compliance Controller
    cartesian_compliance_controller:
      type: cartesian_compliance_controller/CartesianComplianceController


franka_robot_state_broadcaster:
  ros__parameters:
    arm_id: panda

model_example_controller:
  ros__parameters:
    arm_id: panda

joint_trajectory_controller:
  ros__parameters:
    joints:
      - panda_joint1
      - panda_joint2
      - panda_joint3
      - panda_joint4
      - panda_joint5
      - panda_joint6
      - panda_joint7
    command_interfaces:
      - velocity
    state_interfaces:
      - position
      - velocity
    gains:
      panda_joint1: { p: 600., d: 30., i: 0., i_clamp: 1. }
      panda_joint2: { p: 600., d: 30., i: 0., i_clamp: 1. }
      panda_joint3: { p: 600., d: 30., i: 0., i_clamp: 1. }
      panda_joint4: { p: 600., d: 30., i: 0., i_clamp: 1. }
      panda_joint5: { p: 250., d: 10., i: 0., i_clamp: 1. }
      panda_joint6: { p: 150., d: 10., i: 0., i_clamp: 1. }
      panda_joint7: { p: 50., d: 5., i: 0., i_clamp: 1. }

joint_impedance_example_controller:
  ros__parameters:
    arm_id: panda
    k_gains:
      - 24.0
      - 24.0
      - 24.0
      - 24.0
      - 10.0
      - 6.0
      - 2.0
    d_gains:
      - 2.0
      - 2.0
      - 2.0
      - 1.0
      - 1.0
      - 1.0
      - 0.5

cartesian_impedance_controller:
  ros__parameters:
    arm_id: panda
    k_gains:
      - 0.90
      - 0.90
      - 0.90
      - 0.40
      - 0.40
      - 0.40
    d_gains:
      - 1.9
      - 1.9
      - 1.9
      - 1.3
      - 1.3
      - 1.3

move_to_start_example_controller:
  ros__parameters:
    arm_id: panda
    k_gains:
      - 600.0
      - 600.0
      - 600.0
      - 600.0
      - 250.0
      - 150.0
      - 50.0
    d_gains:
      - 30.0
      - 30.0
      - 30.0
      - 30.0
      - 10.0
      - 10.0
      - 5.
joint_position_example_controller:
  ros__parameters:
    arm_id: panda
    k_gains:
      - 24.0
      - 24.0
      - 24.0
      - 24.0
      - 10.0
      - 6.0
      - 2.0
    d_gains:
      - 2.0
      - 2.0
      - 2.0
      - 1.0
      - 1.0
      - 1.0
      - 0.5

joint_velocity_example_controller:
  ros__parameters:
    arm_id: panda
    k_gains:
      - 24.0
      - 24.0
      - 24.0
      - 24.0
      - 10.0
      - 6.0
      - 2.0
    d_gains:
      - 2.0
      - 2.0
      - 2.0
      - 1.0
      - 1.0
      - 1.0
      - 0.5

cartesian_velocity_example_controller:
  ros__parameters:
    arm_id: panda
    k_gains:
      - 24.0
      - 24.0
      - 24.0
      - 24.0
      - 10.0
      - 6.0
      - 2.0
    d_gains:
      - 2.0
      - 2.0
      - 2.0
      - 1.0
      - 1.0
      - 1.0
      - 0.5

# Cartesian Compliance Controller
cartesian_compliance_controller:
  ros__parameters:
      end_effector_link: "panda_link8"
      robot_base_link: "panda_link0"
      ft_sensor_ref_link: "panda_link8"
      compliance_ref_link: "panda_link8"
      joints:
        - panda_joint1
        - panda_joint2
        - panda_joint3
        - panda_joint4
        - panda_joint5
        - panda_joint6
        - panda_joint7
        
      # Choose: position or velocity.
      command_interfaces:
        - velocity
      state_interfaces:
        - position
        - velocity
      stiffness:
        trans_x: 400.0
        trans_y: 400.0
        trans_z: 400.0
        rot_x: 20.0
        rot_y: 20.0
        rot_z: 20.0
      solver:
        error_scale: 0.3
        iterations: 1
      pd_gains:
        trans_x: {p: 0.05, d: 0.005}
        trans_y: {p: 0.05, d: 0.005}
        trans_z: {p: 0.05, d: 0.005}
        rot_x: {p: 0.3}
        rot_y: {p: 0.3}
        rot_z: {p: 0.3}
      gravity:
        x: 0.0
        y: 0.0
        z: -9.81
      tool:
        mass: 1.5
        com_x: 0.0
        com_y: 0.0
        com_z: 0.05 